File: app/src/test/java/info/nightscout/androidaps/utils/JsonHelperTest.java
package info.nightscout.androidaps.utils;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

/**
 * Created by mike on 12.03.2018.
 */

public class JsonHelperTest {

    private String jsonString = "{\"d\":\"3.0\",\"i\":\"4\",\"s\":\"5\",\"b\":\"true\",\"j\":{\"a\": \"1\"}}";

    @Test
    public void safeGetObjectTest() throws JSONException {
        JSONObject object = new JSONObject(jsonString);
        Object o = new Object();
        assertEquals(o, JsonHelper.safeGetObject(null, "x", o));
        assertEquals(o, JsonHelper.safeGetObject(object, "x", o));
        assertNotEquals(o, JsonHelper.safeGetObject(object, "d", o));
    }

    @Test
    public void safeGetJSONObjectTest() throws JSONException {
        JSONObject object = new JSONObject(jsonString);
        JSONObject o = new JSONObject();
        assertEquals(o, JsonHelper.safeGetJSONObject(null, "x", o));
        assertTrue(JsonHelper.safeGetJSONObject(object, "j", o).has("a"));
        assertEquals(o, JsonHelper.safeGetJSONObject(object, "d", o));
    }

    @Test
    public void safeGetStringTest() throws JSONException {
        JSONObject object = new JSONObject(jsonString);
        Object o = new Object();
        assertNull(JsonHelper.safeGetString(null, "s"));
        assertNull(JsonHelper.safeGetString(object, "notexisting"));
        assertEquals("5", JsonHelper.safeGetString(object, "s"));

        assertEquals("default", JsonHelper.safeGetString(null, "notexisting", "default"));
        assertEquals("default", JsonHelper.safeGetString(object, "notexisting", "default"));
        assertEquals("5", JsonHelper.safeGetString(object, "s", "default"));

        assertEquals("default", JsonHelper.safeGetStringAllowNull(null, "notexisting", "default"));
        assertEquals("default", JsonHelper.safeGetStringAllowNull(object, "notexisting", "default"));
        assertNull(JsonHelper.safeGetStringAllowNull(object, "notexisting", null));
        assertEquals("5", JsonHelper.safeGetStringAllowNull(object, "s", "default"));
    }

    @Test
    public void safeGetDoubleTest() throws JSONException {
        JSONObject object = new JSONObject(jsonString);

        assertEquals(0.0d, JsonHelper.safeGetDouble(object, "notexisting"), 0.0d);
        assertEquals(0.0d, JsonHelper.safeGetDouble(null, "notexisting"), 0.0d);
        assertEquals(3.0d, JsonHelper.safeGetDouble(object, "d"), 0.000001d);

        assertEquals(6d, JsonHelper.safeGetDouble(null, "notexisting", 6d), 0.0d);
        assertEquals(6d, JsonHelper.safeGetDouble(object, "notexisting", 6d), 0.0d);
        assertEquals(3d, JsonHelper.safeGetDouble(object, "d", 6d), 0.0d);
    }

    @Test
    public void safeGetLntTest() throws JSONException {
        JSONObject object = new JSONObject(jsonString);
        assertEquals(0, JsonHelper.safeGetInt(null, "notexisting"));
        assertEquals(0, JsonHelper.safeGetInt(object, "notexisting"));
        assertEquals(4, JsonHelper.safeGetInt(object, "i"));
    }

    @Test
    public void safeGetLongTest() throws JSONException {
        JSONObject object = new JSONObject(jsonString);
        assertEquals(0, JsonHelper.safeGetInt(null, "notexisting"));
        assertEquals(0, JsonHelper.safeGetInt(object, "notexisting"));
        assertEquals(4, JsonHelper.safeGetInt(object, "i"));
    }

    @Test
    public void safeGetBooleanTest() throws JSONException {
        JSONObject object = new JSONObject(jsonString);
        assertFalse(JsonHelper.safeGetBoolean(null, "notexisting"));
        assertFalse(JsonHelper.safeGetBoolean(object, "notexisting"));
        assertTrue(JsonHelper.safeGetBoolean(object, "b"));
    }
}

File: core/core-main/src/test/java/info/nightscout/androidaps/utils/JsonHelperTest.kt
package info.nightscout.androidaps.utils

import info.nightscout.interfaces.utils.JsonHelper
import org.json.JSONObject
import org.junit.Assert
import org.junit.jupiter.api.Test

@Suppress("SpellCheckingInspection")
class JsonHelperTest {

    private val jsonString = "{\"d\":\"3.0\",\"i\":\"4\",\"s\":\"5\",\"b\":\"true\",\"j\":{\"a\": \"1\"}}"

    @Test
    fun safeGetObjectTest() {
        val json = JSONObject(jsonString)
        val o = Any()
        Assert.assertEquals(o, JsonHelper.safeGetObject(null, "x", o))
        Assert.assertEquals(o, JsonHelper.safeGetObject(json, "x", o))
        Assert.assertNotEquals(o, JsonHelper.safeGetObject(json, "d", o))
    }

    @Test
    fun safeGetJSONObjectTest() {
        val json = JSONObject(jsonString)
        val o = JSONObject()
        Assert.assertEquals(o, JsonHelper.safeGetJSONObject(null, "x", o))
        Assert.assertTrue(JsonHelper.safeGetJSONObject(json, "j", o)!!.has("a"))
        Assert.assertEquals(o, JsonHelper.safeGetJSONObject(json, "d", o))
    }

    @Test
    fun safeGetStringTest() {
        val json = JSONObject(jsonString)
        Assert.assertNull(JsonHelper.safeGetString(null, "s"))
        Assert.assertNull(JsonHelper.safeGetString(json, "notexisting"))
        Assert.assertEquals("5", JsonHelper.safeGetString(json, "s"))
        Assert.assertEquals("default", JsonHelper.safeGetString(null, "notexisting", "default"))
        Assert.assertEquals("default", JsonHelper.safeGetString(json, "notexisting", "default"))
        Assert.assertEquals("5", JsonHelper.safeGetString(json, "s", "default"))
        Assert.assertEquals("default", JsonHelper.safeGetStringAllowNull(null, "notexisting", "default"))
        Assert.assertEquals("default", JsonHelper.safeGetStringAllowNull(json, "notexisting", "default"))
        Assert.assertNull(JsonHelper.safeGetStringAllowNull(json, "notexisting", null))
        Assert.assertEquals("5", JsonHelper.safeGetStringAllowNull(json, "s", "default"))
    }

    @Test
    fun safeGetDoubleTest() {
        val json = JSONObject(jsonString)
        Assert.assertEquals(0.0, JsonHelper.safeGetDouble(json, "notexisting"), 0.0)
        Assert.assertEquals(0.0, JsonHelper.safeGetDouble(null, "notexisting"), 0.0)
        Assert.assertEquals(3.0, JsonHelper.safeGetDouble(json, "d"), 0.000001)
        Assert.assertEquals(6.0, JsonHelper.safeGetDouble(null, "notexisting", 6.0), 0.0)
        Assert.assertEquals(6.0, JsonHelper.safeGetDouble(json, "notexisting", 6.0), 0.0)
        Assert.assertEquals(3.0, JsonHelper.safeGetDouble(json, "d", 6.0), 0.0)
    }

    @Test
    fun safeGetLntTest() {
        val json = JSONObject(jsonString)
        Assert.assertEquals(0, JsonHelper.safeGetInt(null, "notexisting").toLong())
        Assert.assertEquals(0, JsonHelper.safeGetInt(json, "notexisting").toLong())
        Assert.assertEquals(4, JsonHelper.safeGetInt(json, "i").toLong())
    }

    @Test
    fun safeGetLongTest() {
        val json = JSONObject(jsonString)
        Assert.assertEquals(0, JsonHelper.safeGetInt(null, "notexisting").toLong())
        Assert.assertEquals(0, JsonHelper.safeGetInt(json, "notexisting").toLong())
        Assert.assertEquals(4, JsonHelper.safeGetInt(json, "i").toLong())
    }

    @Test
    fun safeGetBooleanTest() {
        val json = JSONObject(jsonString)
        Assert.assertFalse(JsonHelper.safeGetBoolean(null, "notexisting"))
        Assert.assertFalse(JsonHelper.safeGetBoolean(json, "notexisting"))
        Assert.assertTrue(JsonHelper.safeGetBoolean(json, "b"))
    }
}